staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go

```diff
@@ -113,71 +114,132 @@ func WithPriorityAndFairness(
 				waitingMark.recordReadOnly(int(atomic.AddInt32(&atomicReadOnlyWaiting, delta)))
 			}
 		}
-		var resultCh chan interface{}
-		var forgetWatch utilflowcontrol.ForgetWatchFunc
-		if isWatchRequest {
-			resultCh = make(chan interface{})
+		queueNote := func(inQueue bool) {
+			if inQueue {
+				noteWaitingDelta(1)
+			} else {
+				noteWaitingDelta(-1)
+			}
 		}
-		execute := func() {
-			noteExecutingDelta(1)
-			defer noteExecutingDelta(-1)
-			served = true
 
-			innerCtx := ctx
-			innerReq := r
+		// find the estimated "width" of the request
+		// TODO: Maybe just make it costEstimator and let it return additionalLatency too for the watch?
+		// TODO: Estimate cost should also take fcIfc.GetWatchCount(requestInfo) as a parameter.
+		width := widthEstimator.EstimateWidth(r)
+		digest := utilflowcontrol.RequestDigest{RequestInfo: requestInfo, User: user, Width: width}
 
+		if isWatchRequest {
+			shouldStartWatchCh := make(chan struct{})
 			var watchInitializationSignal utilflowcontrol.InitializationSignal
-			if isWatchRequest {
+			var watchReq *http.Request
+			var forgetWatch utilflowcontrol.ForgetWatchFunc
```

Comment:
isn't it racy in a timeout case to assign to watchInitializationSignal and forgetWatch inside a goroutine, and then use these vars later in this file (in the deferred `Protect from the situation when request will not reach storage layer` and `Forget the watcher if it was registered.` case)?

Reply:
I don't see this race.

The reason is that this branch of code is only for watch requests (which is long-running).
And for long-running requests, the timeout handler is simply delegating and not wrapping anything:

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L84-L86

Reply:
[The other argument is that all the channel-based (or context-based) timeout handlers won't play here.
I mean, the defer function may be called only if we finished one of the:
(a) L215: case <-shouldStartWatchCh:
(b) case err := <-resultCh:

(a) may happen only after the channel was closed which is happening after the corresponding vars are already set.

(b) happens only if the goroutine above finishes, which means the Handle() has also been finished (which is the only thing that calls execute).

So there isn't a race here.